Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1378 @@
+We can refer to either as **derivation**.
+
+We start with a special symbol called the start non-terminal. Then, we find and replace symbols according to our substitution rules. Once we have replaced everything we could have replaced, we stop.
+
+The symbols to be replaced are called **non-terminals**, and the symbols they are ultimately replaced by are called **terminals**. Non-terminals denote things that do not occur in the language itself. Terminals are usually the lexical items of the language – you can think of them being the "words" of the language. They are called so because derivation terminates on them.
+
+The rules for substituting symbols are called **production rules**.
+They are the rules that define the grammar (and, by extension, the language).
+In formal language theory, we usually write them as `alpha -> beta`, where `alpha` is called the production head, and `beta` the production body.
+In each step of derivation, some head is replaced by its corresponding body.
+
+### Bottom-up parsing
+
+Parsing is, in a sense, a process inverse to derivation. Instead of building an arbitrary sentence, we're given a sentence and tasked with recovering the derivation that produced it. 
+
+There are two broad approaches to accomplish that. We could try to run the derivation process forwards, but at each step choose the substitution that would get us closer to the input string. This is top-down parsing. Not all grammars work well with this approach, but it is arguably more straightforward. 
+
+Or we could run the derivation process in reverse by finding a production rule whose body matches some of the input string. We do that until we're back at the starting non-terminal. This is bottom-up parsing. It works well with more grammars (still not all of them, though), and that's the approach Happy adheres to.
+
+In our case, we want to generate a bottom-up parser that will build an **abstract syntax tree** (AST), which is a data structure that describes the exact structure of the language we're parsing.
+
+For an input like `-42 * f x`, we'd get a tree that looks like this:
+
+```haskell
+EBinOp _ (ENeg _ (EInt _ 42)) (Times _) (EApp _ (EVar _ "f") (EVar "x"))
+```
+
+It can be better visualized like this:
+
+```haskell
+                          EBinOp _
+                         /   |    \
+                  ENeg _     |     EApp _
+                 /       Times _  /      \
+        EInt _ 42                /        \
+                       EVar _ "f"          EVar _ "x"
+```
+
+So that Happy could generate the parser, we need to provide a formal grammar in terms of terminals, non-terminals, and production rules. 
+
+The terminals are going to be the tokens that were produced by Alex, and they will correspond to the leaves of the abstract syntax tree (`EInt`, `Times`, and `EVar` in the example above).
+
+This is how we can create terminals in Happy:
```

Comment:
Since terminals are Alex tokens, I believe it is wrong to say that this is how we create them in Happy: we're just aliasing them here. 

I would just say "This is how our terminals look in Happy:"